File: android/src/main/java/viska/android/InstanceActivity.java
package viska.android;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.Database;
import com.couchbase.lite.ListenerToken;
import java.util.ArrayList;
import viska.database.Profile;
import viska.database.ProfileKt;

public abstract class InstanceActivity extends AppCompatActivity {

  protected Profile profile;
  protected Database db;
  private final ArrayList<ListenerToken> tokens = new ArrayList<>();

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    profile = ProfileKt.getProfile(this);
    if (profile == null) {
      Log.i(getClass().getSimpleName(), "No active account, switching to NewProfileActivity");
      moveToNewProfileActivity();
    }

    GlobalState.INSTANCE
        .getCreatingAccount()
        .observe(
            this,
            creatingAccount -> {
              if (creatingAccount) {
                finish();
              }
            });
  }

  @Override
  protected void onStart() {
    super.onStart();

    db = profile.openDatabase();
    startForegroundService(new Intent(this, DaemonService.class));
  }

  @Override
  protected void onStop() {
    super.onStop();

    if (db != null) {
      synchronized (tokens) {
        tokens.forEach(db::removeChangeListener);
        tokens.clear();
      }
      try {
        db.close();
      } catch (CouchbaseLiteException e) {
        Log.e(this.getClass().getCanonicalName(), "Failed to close database", e);
      }
      db = null;
    }
  }

  protected void moveToNewProfileActivity() {
    startActivity(new Intent(this, NewProfileActivity.class));
    finish();
  }

  protected void storeListenerToken(final ListenerToken token) {
    synchronized (tokens) {
      tokens.add(token);
    }
  }
}

File: android/src/main/java/viska/android/InstanceActivity.kt
package viska.android

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlin.IllegalStateException
import viska.daemon.DaemonService
import viska.database.ProfileService

@AndroidEntryPoint
abstract class InstanceActivity : AppCompatActivity() {

  @Inject lateinit var profileService: ProfileService
  @Inject lateinit var daemonService: DaemonService

  private var exiting = false

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (GlobalState.creatingAccount.value) {
      finish()
      throw IllegalStateException(
          "Launching an InstanceActivity is forbidden during account creation")
    }

    if (exiting) {
      finish()
      throw IllegalStateException("Application is exiting")
    }

    synchronized(INSTANCES) { INSTANCES.add(this) }
  }

  override fun onDestroy() {
    synchronized(INSTANCES) {
      INSTANCES.remove(this)
      if (exiting && INSTANCES.isEmpty()) {
        daemonService.stop()
      }
    }
    super.onDestroy()
  }

  protected fun exitApp() {
    exiting = true
    finishAll()
  }

  protected fun redirectToNewProfile() {
    startActivity(Intent(this, NewProfileActivity::class.java))
    finish()
  }

  companion object {
    private val INSTANCES = mutableSetOf<InstanceActivity>()

    fun finishAll() {
      synchronized(INSTANCES) { INSTANCES.forEach { activity -> activity.finish() } }
    }
  }
}
